docs/rfcs/0004-rfc-connectiomanager.md

```diff
@@ -0,0 +1,80 @@
+- Status: In progress
+- Date: 2023-11-02
+- Author: [@gela](https://github.com/gela)
+- RFC PR: TBD
+- Implementation GitHub issue: TBD.
+
+
+## Summary
+
+The existing `ConnectionManager` component will handle all connectivity related checks once during VPN activation (`Controller::StateConnecting`) and then periodically once the VPN is already activated (`Controller::StateOn`); the controller will be the entity which manages the controller states that the client goes through such as `connecting` and `confirming` as we activate and deactivate the VPN to reflect those to the user what is happening behind the scene.
+
+## Motivation
+
+At present, the handling of VPN connectivity lacks consistency for users. There are instances when the VPN goes into No Signal due to a variety of connectivity-related issues, such as the network connection failing or the server location becoming unavailable. In these situations, users are often left without a clear explanation of what went wrong or guidance on how to troubleshoot and resolve the issue. They may encounter a background error message or lose signal without further information.
+
+The objective with the Connection Manager is to address these issues by running probes on all potential causes and presenting the user with an appropriate modal detailing the root cause of their problem. This approach aims to enhance the user's understanding of the issue and provide them with actionable steps to rectify it.
```

Comment:
> "modal"

Does this mean a modal dialog box? Isn't this error displayed using an alert, which is not modal?

Reply:
The errors would be shown to the user using a box that pops up in the client, [this Figma](https://www.figma.com/file/GZKigVZbnCisMJNU3Hc9xM/Connection-error-messaging?type=design&node-id=26-215&mode=design&t=kfYnUKuXQ9nsGDRR-0) shows what they would look like. User would need to explicitly dismiss the alert. I think that's a modal, right? If I'm using the wrong term I can change it to something more accurate.

<img width="393" alt="Screenshot 2023-11-14 at 4 14 44 PM" src="https://github.com/mozilla-mobile/mozilla-vpn-client/assets/3746552/58b41213-d6f9-44c2-8657-a74d12257a90">

Reply:
Looks good, since it is a modal. I incorrectly assumed it was a modeless alert.

Reply:
> Will this modal be automatically closed if the error is transitory and is resolved another change in the system or network? For example, what happens if this occurs while the client app is hidden and the VPN hits a transient connection error and then recovers?

Yes if the issue resolves itself and VPN goes back to stable after we pop up the modal then the modal will close automatically. I'll add this to the document. 

If VPN is in the background as this happens we will still show the same modal and handle it as if it was in the foreground, and if the error resolves itself we will close the modal automatically. @MattLichtenstein I think that's doable, right?